src/ngraph/node.hpp

```diff
@@ -324,6 +376,309 @@ namespace ngraph
         size_t m_placement_index = placement_invalid;
     };
 
+    /// \brief A handle for one of a node's inputs.
+    template <typename NodeType>
+    class Input
+    {
+    public:
+        /// \brief Constructs a Input.
+        /// \param node Pointer to the node for the input handle.
+        /// \param index The index of the input.
+        Input(NodeType* node, size_t index)
+            : m_node(node)
+            , m_index(index)
+        {
+        }
+
+        /// \return A pointer to the node referenced by this input handle.
+        NodeType* get_node() const { return m_node; }
+        /// \return The index of the input referred to by this input handle.
+        size_t get_index() const { return m_index; }
+        /// \return The element type of the input referred to by this input handle.
+        const element::Type& get_element_type() const
+        {
+            return m_node->get_input_element_type(m_index);
+        }
+        /// \return The shape of the input referred to by this input handle.
+        const Shape& get_shape() const { return m_node->get_input_shape(m_index); }
+        /// \return The partial shape of the input referred to by this input handle.
+        const PartialShape& get_partial_shape() const
+        {
+            return m_node->get_input_partial_shape(m_index);
+        }
+        /// \return A handle to the output that is connected to this input.
+        Output<Node> get_source_output() const;
+        /// \return A reference to the tensor descriptor for this input.
+        descriptor::Tensor& get_tensor() const
+        {
+            return m_node->m_inputs.at(m_index).get_output().get_tensor();
```

Comment:
Implement these by calling proper Node methods to get an output. Might not even need to be friends of Node, because who really needs friends?
This holds for next couple of methods, too.
If you don't want to use methods than at least select between .at(x) and [x] in these methods.

Reply:
Doing it this way makes me sad, but:
- We still need the `descriptor::*` classes to be around for now, because existing deprecated APIs return pointers to them.
- We need a single source of truth for input/output info, and for the moment I think that has to be the (legacy) descriptor classes.
- The accessors are deprecated, so if we use them here we would be throwing deprecation warnings in core code (which I was trying hard to get rid of in this PR).
- Once we actually retire the `descriptor::*` accessors, we can come up with a cleaner interface here.

Reply:
Think I've addressed the `at` vs `[]` thing, BTW.